### PR TITLE
Item Group API

### PR DESCRIPTION
**Pull request by modmuss50**
TODO:

- [x] Better button texture/location/size

- [x] Docs

- [x] Expand FabricItemGroup, move more of it to an api/impl 

The idea for this early PR is to get inital feedback and suggestions. I have tried to keep the mixins to a minimum and I feel like I have succeed well. There are a few magic numbers around the place that might need to be changed.

Do we need a way to remove / replace item groups?

The gui is untouched if no item groups are added.

Thanks for your feedback, hopefully this PR can move a bit faster than the container one did.

GIF:

### CODE REVIEW

File: src/main/java/net/fabricmc/fabric/mixin/client/itemgroup/MixinCreativePlayerInventoryGui.java [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+	}
+
+	@Inject(method = "method_2468", at = @At("HEAD"), cancellable = true)
+	private void method_2468(ItemGroup itemGroup, CallbackInfo info){
+		if(!isGroupVisable(itemGroup)){
+			info.cancel();
+		}
+	}
+
+	private boolean isGroupVisable(ItemGroup itemGroup){
```
Comment (falkreon) [inline src/main/java/net/fabricmc/fabric/mixin/client/itemgroup/MixinCreativePlayerInventoryGui.java]:
```suggestion
 	private boolean isGroupVisible(ItemGroup itemGroup){
```

File: src/main/java/net/fabricmc/fabric/api/client/itemgroup/FabricItemGroup.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+	public FabricItemGroup(Identifier identifier) {
+		this.identifier = identifier;
+	}
+
+	public FabricItemGroup icon(Supplier<ItemStack> stackSupplier){
+		this.stackSupplier = stackSupplier;
+		return this;
+	}
+
+	public ItemGroup create(){
```
Comment (asiekierka) [inline src/main/java/net/fabricmc/fabric/api/client/itemgroup/FabricItemGroup.java]: This is a builder, just call it FabricItemGroupBuilder, add a create(Identifier) and rename this to build if so.

File: src/main/java/net/fabricmc/fabric/client/itemgroup/FabricItemGroupUtils.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
+package net.fabricmc.fabric.client.itemgroup;
+
+import net.minecraft.client.gui.widget.ButtonWidget;
+
+import java.util.function.Consumer;
+
+//TODO is utils the best name for this?
+public class FabricItemGroupUtils {
```
Comment (asiekierka) [inline src/main/java/net/fabricmc/fabric/client/itemgroup/FabricItemGroupUtils.java]: FabricCreativeGuiComponents

File: src/main/java/net/fabricmc/fabric/mixin/client/itemgroup/MixinCreativePlayerInventoryGui.java [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+		if(itemGroup == ItemGroup.SEARCH || itemGroup == ItemGroup.INVENTORY){
```
Comment (asiekierka) [inline src/main/java/net/fabricmc/fabric/mixin/client/itemgroup/MixinCreativePlayerInventoryGui.java]: Don't forget to add the hotbar here, and perhaps make this a Set of sorts (even if hardcoded for now!) then replace "/ 10" below with "/ (12 - set.size())".

File: src/main/java/net/fabricmc/fabric/mixin/client/itemgroup/MixinItemGroup.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+	@Shadow @Final @Mutable
+	public static ItemGroup[] GROUPS;
+
+	@Shadow public abstract int getId();
+
+	@Override
+	public void fabric_expandArray() {
+		ItemGroup[] tempGroups = GROUPS;
+		GROUPS = new ItemGroup[GROUPS.length + 1];
+		for(ItemGroup group : tempGroups){
```
Comment (asiekierka) [inline src/main/java/net/fabricmc/fabric/mixin/client/itemgroup/MixinItemGroup.java]: We could probably use System.arraycopy here, but at this size it doesn't matter.

**Comment by modmuss50** [comment]
Here is an updated screenshot of the new button textures in game.

java_2018-12-24_12-00-31


I think this is feature complete now, let me know of any other changes/ideas you have.

**Comment by Jacky2611** [comment]
Haven't been able to look at everything since I am on my phones but what I have seen so far looks promising.
Can't really give you any good feedback on the Java side of things (only recently started working with Java again) but I do have some experience dealing with UI/UX stuff. Tbh I think the buttons are a bit too small. Sure they look clean, but hitting them accurately could be tricky if you are moving fast. And once you have a few mods installed switching between tabs will involve a lot of annoying clicking. I would maybe have tried to change the creative menu to something more Skyrim like (scrollable list of tabs with icon to the left) or something entirely new instead of trying to add to a menu that was clearly never meant to be expanded.
I know that this is not really something important and probably also highly subjective, but whatever gets added to fabric now will probably become the standart and stick around for quite a while.

**Comment by Vexatos** [comment]
One idea would be allowing using the scrollwheel while the cursor is above the tabs to perform the action of the buttons.

**Comment by Jacky2611** [comment]
You would still have to get you cursor above two very small buttons and then you would have to scroll without knowing what tabs you will get on the next page. After each time scrolling you would have to check all the loaded tabs (which are all over the screen) to make sure that you haven't arrived at your target yet. A better UI would have all tabs in a list that shows the current tab and also displays the next and previous tabs so users know when to stop scrolling.
I do have some design ideas that I could draw mockups for, but honestly I am not even sure what you guys are thinking about a more radical UI redesign. Making a somewhat decent UI usually takes quite some time and testing. Not sure if anyone would be willing to invest that much time.

**Comment by asiekierka** [comment]
>I do have some design ideas that I could draw mockups for, but honestly I am not even sure what you guys are thinking about a more radical UI redesign. Making a somewhat decent UI usually takes quite some time and testing. Not sure if anyone would be willing to invest that much time.

Separate mod. Fabric's builtin implementation should be minimal and functional

**Comment by modmuss50** [comment]
The idea of this is to change as little as possible, while having the basic functionality. Other people have expressed interest in replacing the whole gui or large parts of it. This is out of scope for what I am trying to do here.

I think what might be best if a separate mod is made that has free reign over the creative gui. This way it can add scroll bars and new sections and what not.

This PR is very minimal in the amount of code so maintaining it, it will be a lot easier.

**Comment by Jacky2611** [comment]
Should it though? This will be what most if not all users will see when they use fabric. Most users will probably be annoyed that the creative menu is annoying to use but at the same time too lazy to search for a mod that will fix it. I can understand that you want to keep fabric small, but If we are already adding an implementation of custom tags to fabric we might as well do it right. Or alternatively make it easy to add some useful utility mods during the fabric installation as additional features.

**Comment by asiekierka** [comment]
@Jacky2611 
If Fabric API bloats up, it takes longer to port, and if we want to port it once a week during development, preferably within a few hours of a snapshot release... I hope you see the point. The more things we have to maintain, the more work it takes - which is why non-essential things are split away.

> Or alternatively make it easy to add some useful utility mods during the fabric installation as additional features.

That's a better approach and I personally think it's worth considering. A template modpack, for instance?

**Comment by Pannoniae** [comment]
I would +1 the template pack.

**Comment by Jacky2611** [comment]
I was more thinking along the lines of some checkboxes in the installer of the current stable release.
So that stuff like the mod menu, a more mod friendly UI and one or two other utility mods will be suggested and installed per default.
Alternatively I heard rumors of a dependency downloader on the discord server a few days ago. Should that actually become a thing and someone were to write some UI for it to turn it into a mod browser it could be displayed at the initial startup with a list of suggested mods.

Tbh I already had an inkling that my suggestion would be out of scope, but I knew that not mentioning it now and knowing that I had the chance would drive me crazy every time I had to open the creative menu to get my items during testing.

**Comment by asiekierka** [comment]
>dependency downloader

Security concerns. No.

The installer for vanilla profiles is an alternative; since we already build MultiMC instances, I think the preferred approach would simply be for someone in the community to build a "sane defaults" MultiMC instance that also checks if all the versions are stable etc.

File: src/main/java/net/fabricmc/fabric/mixin/client/itemgroup/MixinItemGroup.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+		GROUPS = new ItemGroup[GROUPS.length + 1];
+		for (ItemGroup group : tempGroups) {
+			GROUPS[group.getId()] = group;
+		}
+	}
+
+	@Inject(method = "isTopRow", cancellable = true, at = @At("HEAD"))
+	private void isTopRow(CallbackInfoReturnable<Boolean> info) {
+		if (getId() > 11) {
+			info.setReturnValue((id - 12) % 9 < 4);
```
Comment (asiekierka) [inline src/main/java/net/fabricmc/fabric/mixin/client/itemgroup/MixinItemGroup.java]: "% 9"? Shouldn't we be using COMMON_GROUPS here?

**Comment by modmuss50** [comment]
I’ll get the page counter in later tonight, and then I’m happy with it.

In the future we could look into expanding to allow searching of each tab, this could be done in a none API breaking way.

**Comment by modmuss50** [comment]
Tooltip + tiny possible issue solved.

**Pull request merged**